draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -816,327 +1115,323 @@ Input: alpha, an octet string to be hashed.
-In particular, this method applies to the Montgomery curves y^2 = x^3 + A\*x^2 + x
-setting B=1.
-
-Preconditions: A Montgomery curve such that A!=0.
-
-Input: alpha, an octet string to be hashed.
+#### Implementation, q=5 (mod 8)
 
-Constants: A and B, the parameters of the curve; N, a non-square in F.
-
-Output: (x, y), a point on E.
-
-Operations:
-
-~~~
-1.   u = hash2base(alpha)
-2.  x1 = -A / (1 + N * u^2)
-3. gx1 = x1^3 + A * x1^2 + B * x1
-4.  x2 = -x1 - A
-5. gx2 = x2^3 + A * x2^2 + B * x2
-6.   e = gx1^((q - 1) / 2)
-7. If is_square(gx1), set x = x1 and y = -e * sqrt(gx1)
-8. If is_square(gx2), set x = x2 and y = -e * sqrt(gx2)
-9. Output h*(x, y)
-~~~
-
-#### Implementation
-
-The following procedure implements elligator2 algorithm in a straight-line
-fashion.
+The following is a straight-line implementation of Elligator 2
+for curves where q=5 (mod 8), including Curve25519.
 
 ~~~
-map2curve_elligator2(alpha)
+map2curve_elligator2_5mod8(alpha)
 Input: alpha, an octet string to be hashed.
 Output: (x, y), a point on E.
 
-Constants:  
-1. c1 is an non-square in F.
-2. c2 = (q - 1) / 2      // Integer arithmetic
+Constants:
+1. c1 is Z^((q + 3) / 8) in F.
+2. c2 is sqrt(-1) in F.
 
 Steps:
 1.    u = hash2base(alpha)
-2.   x1 = u^2
-3.   x1 = c1 * x1        
-4.   x1 = x1 + 1
-5.   x1 = 1 / x1
-6.   x1 = A * x1
-7.   x1 = -x1            // x1 = -A / (1 + N * u^2)
-8.  gx1 = x1 + A
-9.  gx1 = gx1 * x1
-10. gx1 = gx1 + B
-11. gx1 = gx1 * x1       // gx1 = x1^3 + A * x1^2 + B * x1
-12.  x2 = -x1 - A        // x2 = -x1 - A
-13. gx2 = x2 + A
-14. gx2 = gx2 * x2
-15. gx2 = gx2 + B
-16. gx2 = gx2 * x2       // gx2 = x2^3 + A * x2^2 + B * x2
-17.   e = is_square(gx1, q)
-18.   x = CMOV(x2, x1, e)    // If e=True, x=x1, else x=x2
-19.  gx = CMOV(gx2, gx1, e)  // If e=True, x=gx1, else x=gx2
-20.  y2 = sqrt(gx, q)
-22.  y1 = -y2
-22.   y = CMOV(y2, y1, e)    // If e=True, y=y1, else y=y2
-23. Output h*(x, y)
+2.   t1 = u^2
+3.   t1 = Z * t1
+4.   x1 = t1 + 1
+5.   x1 = inv0(x1)            // cannot be 0 because q=5 mod 8
+6.   x1 = -A * x1             // x1 = -A / (1 + Z * u^2)
+7.  gx1 = x1 + A
+8.  gx1 = gx1 * x1
+9.  gx1 = gx1 + B
+10. gx1 = gx1 * x1            // gx1 = x1^3 + A * x1^2 + B * x1
+11. y11 = gx1^((q + 3) / 8)
+12. y12 = c2 * y11
+13.  e1 = y12^2 == gx1
+14.  y1 = CMOV(y11, y12, e1)  // if gx1 is square, this is its sqrt
+15.  x2 = -x1 - A
+16. y21 = y11 * u
+17. y21 = y21 * c1
+18. y22 = c2 * y21
+19. gx2 = t1 * gx1
+20.  e2 = y22^2 == gx2
+21.  y2 = CMOV(y21, y22, e2)  // if gx2 is square, this is its sqrt
+22.  e3 = y1^2 == gx1
+23.   x = CMOV(x2, x1, e3)    // if e == True, x=x1, else x=x2
+24.   y = CMOV(y2, y1, e3)    // if e == True, y=y1, else y=y2
+25.  e4 = sgn0(u) == sgn0(y)  // fix sign of y
+26.   y = CMOV(-y, y, e4)
+27. Output h * (x, y)
 ~~~
 
 ## Encodings for twisted Edwards curves
 
-### Elligator2 Method {#ell2edwards}
+### Elligator 2 Method {#ell2edwards}
```

Comment:
The reason I suggest making this change is because the prior version was incredibly inefficient---it requires something like five exponentiations. This one requires at most two, and that can be reduced to one if implemented in projective coordinates.

Beyond that, using the same mapping for Montgomery and Edwards means that the same implementation can be used for multiple purposes, which is a pretty clear win from a practical perspective.